### PR TITLE
[Contracts/Service] add LazyString with new fromStringable(), isStringable() and resolve() methods

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | -
| Tickets       | -
| License       | MIT
| Doc PR        | -

Its purpose is so generic that I'd like to move `LazyString` from DI to the `Service` contracts.

This PR also adds three new methods to the class, for very common boilerplate we have to write again and again otherwise:
- `LazyString::isStringable($value)` -> ensure a value can be safely cast to string, considering `__toString()` too
- `LazyString::fromStringable($value)` -> turns any object with `__toString()` into a `LazyString`
- `LazyString::resolve($value)` -> casts a value to a string or throw a `TypeError` if not possible (`strval()` only throws a notice, that's why we need this one.)

### CODE REVIEW

File: src/Symfony/Contracts/Service/LazyString.php [hunk 1]
```diff
@@ -49,6 +49,47 @@ public static function fromCallable($callback, ...$arguments): self
...
+    }
+
+    /**
+     * Casts scalars and stringable objects to strings.
+     *
+     * @param object|string|int|float|bool $value
+     *
+     * @throws \TypeError When the provided value is not stringable
+     */
+    final public static function resolve($value): string
```
Comment (yceruto) [inline src/Symfony/Contracts/Service/LazyString.php]: What would be the advantage of using this method over `(string) $value` directly?

Comment (nicolas-grekas) [reply src/Symfony/Contracts/Service/LazyString.php]:
```
echo (string) [];
```
> Notice: Array to string conversion

Comment (yceruto) [reply src/Symfony/Contracts/Service/LazyString.php]: Got it, thanks!

**Review by stof — CHANGES_REQUESTED** [review]
Tests should be added for the new utilities.

Btw, does `isStringable` deal properly with objects having a non-public `__toString` method ?

File: src/Symfony/Contracts/Service/LazyString.php [hunk 2]
```diff
@@ -49,6 +49,47 @@ public static function fromCallable($callback, ...$arguments): self
...
+     */
+    final public static function isStringable($value): bool
+    {
+        return \is_string($value) || $value instanceof self || (\is_object($value) && method_exists($value, '__toString')) || is_scalar($value);
+    }
+
+    /**
+     * Casts scalars and stringable objects to strings.
+     *
+     * @param object|string|int|float|bool $value
```
Comment (stof) [inline src/Symfony/Contracts/Service/LazyString.php]: should `null` be supported too ?

Comment (nicolas-grekas) [reply src/Symfony/Contracts/Service/LazyString.php]: I don't think so - if one wants to deal with `null`, one can always do `resolve($foo ?? 'bar')`

**Comment by nicolas-grekas** [comment]
> Tests should be added for the new utilities.

tests added

> does isStringable deal properly with objects having a non-public __toString method ?

now yes! - another boilerplate that is missing from similar checks in the code btw.

**Comment by nicolas-grekas** [comment]
Status: needs review

**Review by stof — APPROVED** [review]

**Review by yceruto — APPROVED** [review]

**Pull request closed without merging**